[Quant][Inductor] Enable quantized conv weight prepack inside inductor constant folding

leslie-fang-intel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #107951
* #106958
* #106838
* #106836
* #107565
* #105996
* #105906
* #105639
* #105456
* #105455
* #104590
* #104588
* __->__ #104581
* #104580

**Summary**
Enable quantization conv weight prepack inside inductor constant folding.

**Test Plan**
```
python -m pytest test_mkldnn_pattern_matcher.py -k test_qconv2d_unary
```

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

leslie-fang-intel committed:
Enable quantized conv weight prepack inside inductor constant folding

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][Inductor] Enable quantized conv weight prepack inside inductor constant folding"


**Summary**
Enable quantization conv weight prepack inside inductor constant folding.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][Inductor] Enable quantized conv weight prepack inside inductor constant folding"


**Summary**
Enable quantization conv weight prepack inside inductor constant folding.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][Inductor] Enable quantized conv weight prepack inside inductor constant folding"


**Summary**
Enable quantization conv weight prepack inside inductor constant folding.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][Inductor] Enable quantized conv weight prepack inside inductor constant folding"


**Summary**
Enable quantization conv weight prepack inside inductor constant folding.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][Inductor] Enable quantized conv weight prepack inside inductor constant folding"


**Summary**
Enable quantization conv weight prepack inside inductor constant folding.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][Inductor] Enable quantized conv weight prepack inside inductor constant folding"


**Summary**
Enable quantization conv weight prepack inside inductor constant folding.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][Inductor] Enable quantized conv weight prepack inside inductor constant folding"


**Summary**
Enable quantization conv weight prepack inside inductor constant folding.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][Inductor] Enable quantized conv weight prepack inside inductor constant folding"


**Summary**
Enable quantization conv weight prepack inside inductor constant folding.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][Inductor] Enable quantized conv weight prepack inside inductor constant folding"


**Summary**
Enable quantization conv weight prepack inside inductor constant folding.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][Inductor] Enable quantized conv weight prepack inside inductor constant folding"


**Summary**
Enable quantization conv weight prepack inside inductor constant folding.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][Inductor] Enable quantized conv weight prepack inside inductor constant folding"


**Summary**
Enable quantization conv weight prepack inside inductor constant folding.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][Inductor] Enable quantized conv weight prepack inside inductor constant folding"


**Summary**
Enable quantization conv weight prepack inside inductor constant folding.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][Inductor] Enable quantized conv weight prepack inside inductor constant folding"


**Summary**
Enable quantization conv weight prepack inside inductor constant folding.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][Inductor] Enable quantized conv weight prepack inside inductor constant folding"


**Summary**
Enable quantization conv weight prepack inside inductor constant folding.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8

[ghstack-poisoned]

jerryzh168 commented on aten/src/ATen/native/quantized/library.cpp:
```diff
@@ -264,4 +264,5 @@ TORCH_LIBRARY(onednn, m) {
 
   // Conv2D with binary postop
   m.def(TORCH_SELECTIVE_SCHEMA("onednn::qconv2d_pointwise.binary(Tensor qx, float x_scale, int x_zero_point, Tensor qaccum, float accum_scale, int accum_zero_point, Tensor qw, Tensor w_scale, Tensor w_zero_point, Tensor? bias, int[] stride, int[] padding, int[] dilation, int groups, float output_scale, int output_zero_point, bool fp32_output, str binary_attr, Scalar? alpha, str? unary_attr, Scala ...
+  m.def(TORCH_SELECTIVE_SCHEMA("onednn::dynamic_quant_qconv.tensor(Tensor x, float x_scale, int x_zero_point, Tensor prepacked_w, Tensor w_scale, Tensor w_zero_point, int w_axis, Tensor? bias, int[] stride, int[] padding, int[] dilation, bool transposed,  int[] output_padding, int groups) -> Tensor"));
```
nit: change naming to
qconv1d_dynamic, qconv2d_dynamic etc. to be consistent with https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/quantized/library.cpp#L76-L78?

leslie-fang-intel replied:
Thanks for the comment. I am planning to refactor this PR to remove these `dynamic_quant_qconv` op. May ask your kind review again after it.

leslie-fang-intel replied:
Thanks for the comments. I have refactor the code to remove the `dynamic_quant_qconv` op.

jerryzh168 commented on torch/_inductor/freezing.py:
```diff
@@ -123,6 +123,15 @@ def run_node(self, node):
 
         # TODO - remove constant from node_replacement when it has no uses
         if node.op != "get_attr" and isinstance(out, torch.Tensor):
+            if (
```
does constant folding have a way to configure pure/impure ops?

leslie-fang-intel replied:
Sorry, what's the means of pure/impure ops here? Is it something like the black list or white list?

jerryzh168 replied:
pure ops means ops that doesn't have side effect, so they can be constant folded,
impure ops means ops that has side effect and can not be constant folded, I remember Torchscript or maybe fx has a way to mark the op as impure and constant folding will stop at the impure ops

leslie-fang-intel replied:
Great suggestions. I found `impure` is used in fx to eliminate deadcode and generate `FoldedGraphModule`:

- https://github.com/pytorch/pytorch/blob/07107919297db3f8ab37f11c12666b6d6d5f692e/torch/fx/graph.py#L1427
- https://github.com/pytorch/pytorch/blob/07107919297db3f8ab37f11c12666b6d6d5f692e/torch/fx/experimental/const_fold.py#L181

Here, we also add the `is_impure` check which only contains check of `dequantize_per_channel` for now.

jerryzh168 commented on torch/_inductor/fx_passes/quantization.py:
```diff
@@ -142,3 +145,117 @@ def qconv(match: Match, *args, **kwargs):
+        dequant_per_channel = match.nodes[3]
+        conv_node = match.nodes[4]
+
+        print("---- match dequant weight prepack pattern ----", flush=True)
```
nit: remove before landing

leslie-fang-intel replied:
Thanks and removed.

leslie-fang-intel committed:
Update on "[Quant][Inductor] Enable quantized conv weight prepack inside inductor constant folding"


**Summary**
Enable quantization conv weight prepack inside inductor constant folding.

**Test Plan**
```
python -m pytest test_mkldnn_pattern_matcher.py -k test_qconv2d_weight_prepack
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][Inductor] Enable quantized conv weight prepack inside inductor constant folding"


**Summary**
Enable quantization conv weight prepack inside inductor constant folding.

**Test Plan**
```
python -m pytest test_mkldnn_pattern_matcher.py -k test_qconv2d_weight_prepack
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][Inductor] Enable quantized conv weight prepack inside inductor constant folding"


**Summary**
Enable quantization conv weight prepack inside inductor constant folding.

**Test Plan**
```
python -m pytest test_mkldnn_pattern_matcher.py -k test_qconv2d_weight_prepack
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][Inductor] Enable quantized conv weight prepack inside inductor constant folding"


**Summary**
Enable quantization conv weight prepack inside inductor constant folding.

**Test Plan**
```
python -m pytest test_mkldnn_pattern_matcher.py -k test_qconv2d_weight_prepack
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][Inductor] Enable quantized conv weight prepack inside inductor constant folding"


**Summary**
Enable quantization conv weight prepack inside inductor constant folding.

**Test Plan**
```
python -m pytest test_mkldnn_pattern_matcher.py -k test_qconv2d_weight_prepack
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8

[ghstack-poisoned]

leslie-fang-intel commented:
Hi @jgong5 @Xia-Weiwen, I have refactor the quantization weight prepack implementation as synced offline. Please help to take a look.

jgong5 commented on torch/_inductor/fx_passes/quantization.py:
```diff
@@ -142,3 +164,196 @@ def qconv(match: Match, *args, **kwargs):
+    )
+    def qconv_weight_prepack(match: Match, *args, **kwargs):
+        """
+        Macth the pattern:
```
```suggestion
        Match the pattern:
```

leslie-fang-intel replied:
Thanks and changed.

jgong5 commented on torch/_inductor/fx_passes/quantization.py:
```diff
@@ -142,3 +164,196 @@ def qconv(match: Match, *args, **kwargs):
+        Insert weight prepack node and change the pattern to:
+        int8 activation
+          |
+        onednn.qconv2d_pointwise <- onednn.qconv_prepack <- optional(aten.clone.default) <- dequant_per_channel
```
qconv2d accepts an fp32 prepacked weight here?

leslie-fang-intel replied:
Oh, sorry for the mistake of this diagram. Will change it.

leslie-fang-intel replied:
Changed the diagram.

jgong5 commented on torch/_inductor/freezing.py:
```diff
@@ -135,13 +135,35 @@ def run(self):
         return super().run(initial_env=env)
 
 
+def is_impure(node: torch.fx.node.Node):
```
Is it better to do the check in the `ConstantFolder`? Then you only need to check the `dequant` node, not need to worry about other nodes that consume `dequant` output since it is already marked as unknown value.

leslie-fang-intel replied:
Thanks for the suggestions and changed.

kimishpatel commented on torch/_inductor/freezing.py:
```diff
@@ -135,13 +135,35 @@ def run(self):
+        # For the pattern fp32_weight -> quantized_decomposed.quantize_per_channel.default
+        # -> quantized_decomposed.dequantize_per_channel.default
+        # We only folding fp32_weight -> quantized_decomposed.quantize_per_channel.default into
+        # int8_weight and leave quantized_decomposed.dequantize_per_channel.default in graph to be fused
```
@jerryzh168 isnt this something we were planning to do for quant workflow itself? quant workflow should produce a quantized model with folded weights. Can we do that asap so that changes like this are not required?

jerryzh168 replied:
yeah we can do this in quant flow, but ideally we just rely on constant folding the same way in this PR so we don't reinvent the wheel

kimishpatel replied:
sure. whatever the mechanism is, it should be part of the quant workflow is what i meant. quant workflow should give you a model with weights quantized.

jerryzh168 replied:
sure, that sounds good, we can do that after this PR

kimishpatel commented on torch/_inductor/fx_passes/quantization.py:
```diff
@@ -142,3 +164,196 @@ def qconv(match: Match, *args, **kwargs):
+    @register_freezing_graph_pattern(
+        pattern,
+        extra_check=_is_valid_dequant_conv2d_pattern,
+        pass_number=1,  # pass_number=1, ensure it's behand dequant promotion pass
```
```suggestion
        pass_number=1,  # pass_number=1, ensure it's behind dequant promotion pass
```

leslie-fang-intel replied:
Thanks and changed.

kimishpatel commented on torch/_inductor/fx_passes/quantization.py:
```diff
@@ -142,3 +164,196 @@ def qconv(match: Match, *args, **kwargs):
+        ) = match.nodes[0:4]
+
+        clone_node = (
+            match.nodes[4] if has_clone_to_channel_last_node_in_pattern else None
```
are the matches gauranteed to be in this order?

leslie-fang-intel replied:
Thanks for the comment. Only the output node is guaranteed. So change it to
```
conv_node = match.output_node()
mul_node = conv_node.args[0]
sub_node = mul_node.args[0]
to_fp32_node = sub_node.args[0]
```

kimishpatel commented on torch/_inductor/fx_passes/quantization.py:
```diff
@@ -142,3 +164,196 @@ def qconv(match: Match, *args, **kwargs):
+
+        x_shape = qx.meta.get("tensor_meta").shape
+        graph = match.graph
+        with graph.inserting_before(conv_node):
```
I think what we can potentially do is to define `kwargs` or `literals` dictionary that maps kwargs/literals from `pattern` to be found to kwargs/literals from `replacement_pattern`. 

I can imagine an API like:
```
pattern, kwargs = generate_pattern_graph(pattern)
replacement_pattern, kwargs_replacement = generate_pattern_graph(replacement_pattern)
map_from_kwargs_to_replacement_kwargs = generate_mapping(kwargs, kwargs_replacement) # this function is user define where user tells how to map kwargs from pattern to kwargs of replacement_pattern

graph_rewrite(pattern, replacement_pattern, kwargs, map_from_kwargs_to_replacement_kwargs)
```


@angelayi does something like this work?

angelayi replied:
Why do you need to pass in a separate mapping? Why can't it be in the replacement graph?

kimishpatel replied:
replacement graph has those literals burnt in. You dont want to use those values. You want to extract values from matched pattern. That is the whole problem with subgraph rewriter.

kimishpatel commented on torch/_inductor/fx_passes/quantization.py:
```diff
@@ -142,3 +164,196 @@ def qconv(match: Match, *args, **kwargs):
+                padding,
+                dilation,
+                groups,
+                1.0,  # output_scale
+                0,  # output_zero_point
```
This is 1 and 0 because the op being inserted generated fp32 output that gets quantized separately? and later you do appropriate fusion?

leslie-fang-intel replied:
Yean, here we fuse it into the conv which generated fp32 output. Then, we will further check if there is another quant pattern at output in the following passes.

kimishpatel reviewed (commented):
left some comments

jerryzh168 commented on torch/_inductor/fx_passes/quantization.py:
```diff
@@ -142,3 +164,196 @@ def qconv(match: Match, *args, **kwargs):
+
+        # Activation QParams
+        qx, x_zp, x_scale = (
+            kwargs["x"],
+            kwargs["x_zp"],
+            kwargs["x_scale"],
```
yeah I think they are using https://github.com/pytorch/pytorch/blob/f2b4be8a491ced925df8c5f10b49675e1fa0e012/torch/_inductor/fx_passes/quantization.py#L20, this works for simple patterns, but I'm also thinking about the string pattern that allows us to refer to the internal nodes that would be helpful for annotation as well

kimishpatel replied:
Issue is that you dont know what the traced pattern would look like. If you hand write the pattern yourself then you can probably do that but even hand written pattern is just python callable. So it does not directly map to GraphModule.graph.nodes. For that you will have to construct `graph` yourself by inserting `nodes` into the graph that will become your pattern and then you can truly refer to the internal nodes.

jerryzh168 replied:
it is certainly possible to develop some tools to have a map from string name of the node to the node object I think, e.g. :
```
graph_str = "conv = torch.ops.aten.convolution(...)"
map = some_graph_utils(graph_str)
conv_node = map["conv"] 
```

jerryzh168 commented on torch/_inductor/fx_passes/quantization.py:
```diff
@@ -142,3 +164,196 @@ def qconv(match: Match, *args, **kwargs):
     _register_quantized_conv_lowering(quantize_conv_output_pattern)
+
+
+def _is_valid_dequant_conv2d_pattern(match):
```
could you add some comments for this function? what is this doing?

leslie-fang-intel replied:
Thanks and comments added.

jerryzh168 commented on torch/_inductor/fx_passes/quantization.py:
```diff
@@ -142,3 +164,196 @@ def qconv(match: Match, *args, **kwargs):
+            node.target == aten.clone.default for node in match.nodes
+        )
+
+        (
+            to_fp32_node,
+            sub_node,
+            mul_node,
+            dequant_per_channel,
+        ) = match.nodes[0:4]
```
I think if we have a string to node mapping then we can retrieve the node by string name, like what we have in TorchScript, and it will be more robust than this as well

leslie-fang-intel replied:
Yean, we shouldn't write it as this. And I have changed it into
```
conv_node = match.output_node()
mul_node = conv_node.args[0]
sub_node = mul_node.args[0]
to_fp32_node = sub_node.args[0]
```
Since the pattern only has 1 output node as `conv_node`, so these node order are guaranteed. 
Alternately, there is another way to use `filter_nodes` method with node target as 
```
conv_nodes = filter_nodes(match.nodes, aten.convolution.default)
mul_nodes = filter_nodes(match.nodes, aten.mul.Tensor)
sub_nodes = filter_nodes(match.nodes, aten.sub.Tensor)
```
It will return all the nodes with the target op in the matched pattern.

leslie-fang-intel committed:
Update on "[Quant][Inductor] Enable quantized conv weight prepack inside inductor constant folding"


**Summary**
Enable quantization conv weight prepack inside inductor constant folding.

**Test Plan**
```
python -m pytest test_mkldnn_pattern_matcher.py -k test_qconv2d_weight_prepack
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][Inductor] Enable quantized conv weight prepack inside inductor constant folding"


**Summary**
Enable quantization conv weight prepack inside inductor constant folding.

**Test Plan**
```
python -m pytest test_mkldnn_pattern_matcher.py -k test_qconv2d_weight_prepack
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][Inductor] Enable quantized conv weight prepack inside inductor constant folding"


**Summary**
Enable quantization conv weight prepack inside inductor constant folding.

**Test Plan**
```
python -m pytest test_mkldnn_pattern_matcher.py -k test_qconv2d_unary
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8

[ghstack-poisoned]

leslie-fang-intel commented:
Thanks for the comments @jgong5 @jerryzh168 @kimishpatel. Please help to take a look again.

leslie-fang-intel committed:
Update on "[Quant][Inductor] Enable quantized conv weight prepack inside inductor constant folding"


**Summary**
Enable quantization conv weight prepack inside inductor constant folding.

**Test Plan**
```
python -m pytest test_mkldnn_pattern_matcher.py -k test_qconv2d_unary
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][Inductor] Enable quantized conv weight prepack inside inductor constant folding"


**Summary**
Enable quantization conv weight prepack inside inductor constant folding.

**Test Plan**
```
python -m pytest test_mkldnn_pattern_matcher.py -k test_qconv2d_unary
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8

[ghstack-poisoned]

leslie-fang-intel commented:
Hi @eellison @jansel, could you kindly help to take a look of this PR?

eellison commented on torch/_meta_registrations.py:
```diff
@@ -1866,6 +1866,44 @@ def meta_mkl_linear(
+            None,
+        )
+        out = x.new_empty(shape_out, dtype=(torch.float32 if fp32_output else None))
+        out = out.to(memory_format=torch.channels_last)
```
does it always do channels last

leslie-fang-intel replied:
Thanks for the comment. Yes, for the quantization case, we will only use `channels_last` format for convolution.

eellison commented on torch/_inductor/fx_passes/quantization.py:
```diff
@@ -142,3 +164,198 @@ def qconv(match: Match, *args, **kwargs):
+                torch.ops.onednn.qconv2d_pointwise.default, args=new_args
+            )
+            conv_node.replace_all_uses_with(new_conv_node)
+            new_conv_node.meta.update(conv_node.meta)
```
is the meta here exactly the same? or is it a different dtype ?

leslie-fang-intel replied:
Thanks for the comment and good point @eellison. I haven't considerate the `dtype` in node meta too much previously. I think the `dtype` in `conv_node.meta` means the `dtype` of the `conv's output tensor` (not sure if my understanding is right). If so, here the `new_conv_node` will still generate output tensor with float32 dtype, meanwhile it will accept input tensor with uint8 dtype. So, it should be ok to copy the `conv_node.meta` into `new_conv_node`.

eellison replied:
It should be fine in that case, meta just is about the output.

eellison reviewed (approved):
couple questions but looks good

leslie-fang-intel committed:
Update on "[Quant][Inductor] Enable quantized conv weight prepack inside inductor constant folding"


**Summary**
Enable quantization conv weight prepack inside inductor constant folding.

**Test Plan**
```
python -m pytest test_mkldnn_pattern_matcher.py -k test_qconv2d_unary
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][Inductor] Enable quantized conv weight prepack inside inductor constant folding"


**Summary**
Enable quantization conv weight prepack inside inductor constant folding.

**Test Plan**
```
python -m pytest test_mkldnn_pattern_matcher.py -k test_qconv2d_unary
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][Inductor] Enable quantized conv weight prepack inside inductor constant folding"


**Summary**
Enable quantization conv weight prepack inside inductor constant folding.

**Test Plan**
```
python -m pytest test_mkldnn_pattern_matcher.py -k test_qconv2d_unary
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][Inductor] Enable quantized conv weight prepack inside inductor constant folding"


**Summary**
Enable quantization conv weight prepack inside inductor constant folding.

**Test Plan**
```
python -m pytest test_mkldnn_pattern_matcher.py -k test_qconv2d_unary
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][Inductor] Enable quantized conv weight prepack inside inductor constant folding"


**Summary**
Enable quantization conv weight prepack inside inductor constant folding.

**Test Plan**
```
python -m pytest test_mkldnn_pattern_matcher.py -k test_qconv2d_unary
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][Inductor] Enable quantized conv weight prepack inside inductor constant folding"


**Summary**
Enable quantization conv weight prepack inside inductor constant folding.

**Test Plan**
```
python -m pytest test_mkldnn_pattern_matcher.py -k test_qconv2d_unary
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][Inductor] Enable quantized conv weight prepack inside inductor constant folding"


**Summary**
Enable quantization conv weight prepack inside inductor constant folding.

**Test Plan**
```
python -m pytest test_mkldnn_pattern_matcher.py -k test_qconv2d_unary
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][Inductor] Enable quantized conv weight prepack inside inductor constant folding"


**Summary**
Enable quantization conv weight prepack inside inductor constant folding.

**Test Plan**
```
python -m pytest test_mkldnn_pattern_matcher.py -k test_qconv2d_unary
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][Inductor] Enable quantized conv weight prepack inside inductor constant folding"


**Summary**
Enable quantization conv weight prepack inside inductor constant folding.

**Test Plan**
```
python -m pytest test_mkldnn_pattern_matcher.py -k test_qconv2d_unary
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][Inductor] Enable quantized conv weight prepack inside inductor constant folding"


**Summary**
Enable quantization conv weight prepack inside inductor constant folding.

**Test Plan**
```
python -m pytest test_mkldnn_pattern_matcher.py -k test_qconv2d_unary
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][Inductor] Enable quantized conv weight prepack inside inductor constant folding"


**Summary**
Enable quantization conv weight prepack inside inductor constant folding.

**Test Plan**
```
python -m pytest test_mkldnn_pattern_matcher.py -k test_qconv2d_unary
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][Inductor] Enable quantized conv weight prepack inside inductor constant folding"


**Summary**
Enable quantization conv weight prepack inside inductor constant folding.

**Test Plan**
```
python -m pytest test_mkldnn_pattern_matcher.py -k test_qconv2d_unary
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][Inductor] Enable quantized conv weight prepack inside inductor constant folding"


**Summary**
Enable quantization conv weight prepack inside inductor constant folding.

**Test Plan**
```
python -m pytest test_mkldnn_pattern_matcher.py -k test_qconv2d_unary
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][Inductor] Enable quantized conv weight prepack inside inductor constant folding"


**Summary**
Enable quantization conv weight prepack inside inductor constant folding.

**Test Plan**
```
python -m pytest test_mkldnn_pattern_matcher.py -k test_qconv2d_unary
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][Inductor] Enable quantized conv weight prepack inside inductor constant folding"


**Summary**
Enable quantization conv weight prepack inside inductor constant folding.

**Test Plan**
```
python -m pytest test_mkldnn_pattern_matcher.py -k test_qconv2d_unary
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][Inductor] Enable quantized conv weight prepack inside inductor constant folding"


**Summary**
Enable quantization conv weight prepack inside inductor constant folding.

**Test Plan**
```
python -m pytest test_mkldnn_pattern_matcher.py -k test_qconv2d_unary
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][Inductor] Enable quantized conv weight prepack inside inductor constant folding"


**Summary**
Enable quantization conv weight prepack inside inductor constant folding.

**Test Plan**
```
python -m pytest test_mkldnn_pattern_matcher.py -k test_qconv2d_unary
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][Inductor] Enable quantized conv weight prepack inside inductor constant folding"


**Summary**
Enable quantization conv weight prepack inside inductor constant folding.

**Test Plan**
```
python -m pytest test_mkldnn_pattern_matcher.py -k test_qconv2d_unary
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][Inductor] Enable quantized conv weight prepack inside inductor constant folding"


**Summary**
Enable quantization conv weight prepack inside inductor constant folding.

**Test Plan**
```
python -m pytest test_mkldnn_pattern_matcher.py -k test_qconv2d_unary
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][Inductor] Enable quantized conv weight prepack inside inductor constant folding"


**Summary**
Enable quantization conv weight prepack inside inductor constant folding.

**Test Plan**
```
python -m pytest test_mkldnn_pattern_matcher.py -k test_qconv2d_unary
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8

[ghstack-poisoned]

leslie-fang-intel commented:
@pytorchbot merge
